### PR TITLE
add option to load metrics with kwargs

### LINKED ISSUE #687: impossible to pass args to the metrics
**Is your feature request related to a problem? Please describe.**

I'd like to pass arguments to the metric classes in the config file.

[`metric_collection_from_names()`](https://github.com/openvinotoolkit/anomalib/blob/3de3eefc58aee6538a7b2a8b23cf62d287164b1b/anomalib/utils/metrics/__init__.py#L43) instantiates the metric modules by receiving a list of names of metrics, which are extracted from `anomalib.utils.metrics`.

**Describe the solution you'd like**

I'd like something like pytorchlightnings's API for lightningmodules config: 

```
data:
  class_path: anomalib.data.MVTec
  init_args:
    root: ./datasets/MVTec
    category: bottle
    # ...
```

**example of usage**

The "Area Under" kind of metrics don't need to be integrated from `fpr = 0` to `fpr = 1` (`AUPRO` is usually defined from 0 to 0.3 for instance).

I'd like to use that kind of option.

**deprecation of `tools/train.py`**

Will this be useless with the deprecation of `tools/train.py`?
I was guessing that no because `metric_collection_from_names()` is also used in `MetricsConfigurationCallback`.

### PR DESCRIPTION
**Pull request by jpcbertoldo**
# Description

- Fixes #687 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist

- [X] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

**Review by samet-akcay — CHANGES_REQUESTED** [review]
I agree that it would be good to follow the same convention as the PL CLI. 
Only few minor comments.

File: anomalib/utils/metrics/__init__.py [hunk 1]
```diff
@@ -68,3 +68,81 @@ def metric_collection_from_names(metric_names: List[str], prefix: Optional[str])
         else:
             warnings.warn(f"No metric with name {metric_name} found in Anomalib metrics or TorchMetrics.")
     return metrics
+
+
+def _validate_metrics_dict(metrics: Dict[str, Dict[str, Any]]) -> None:
```
[the remark this answers is not in the document]
Comment (jpcbertoldo) [reply anomalib/utils/metrics/__init__.py:73]: I think it ignores private functions, no?

Anyway this gives me the opportunity to ask (more generally): is it a convention to docstring this type of function?

By "this type" i mean "private and trivial".

**Comment by jpcbertoldo** [comment]
Since i'm working on this file, could be nice to solve #685 as well?

**Comment by samet-akcay** [comment]
> Since i'm working on this file, could be nice to solve #685 as well?

Apologies, just seen that issue. Yeah, I agree. It would make sense to move it to test helpers if the tests are the only place of use.

**Comment by jpcbertoldo** [comment]
Both done

**Comment by jpcbertoldo** [comment]
image

I adressed the codacy checks above.

File: anomalib/utils/metrics/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/utils/metrics/__init__.py:56]: > I think it ignores private functions, no?

You are right, it apparently ignores private functions.

For me the most important thing is the clarity. As long as the function is simple, readable and easy to follow, than it is fine.

**Comment by jpcbertoldo** [comment]
Should this go to docs or can we just assume people will guess it because it's consistent with other parts of the config? :upside_down_face:

**Comment by samet-akcay** [comment]
> Should this go to docs or can we just assume people will guess it because it's consistent with other parts of the config? 🙃

would be good to document it somewhere; otherwise, it might stay as a hidden gem :)

Not sure about where to document it though. @ashwinvaidya17, any thoughts?

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for the efforts. This is a good addition to metrics configuration flexibility. I have a few comments though.

File: anomalib/utils/metrics/__init__.py [hunk 2]
```diff
@@ -23,23 +23,6 @@
 __all__ = ["AUROC", "AUPR", "AUPRO", "OptimalF1", "AnomalyScoreThreshold", "AnomalyScoreDistribution", "MinMax", "PRO"]
 
 
-def get_metrics(config: Union[ListConfig, DictConfig]) -> Tuple[AnomalibMetricCollection, AnomalibMetricCollection]:
```
[the remark this answers is not in the document]
Comment (jpcbertoldo) [reply anomalib/utils/metrics/__init__.py:26]: i moved it to test helpers, looks good?

**Comment by jpcbertoldo** [comment]
I believe i addressed everything.

**Comment by samet-akcay** [comment]
> I believe i addressed everything.

Thanks @jpcbertoldo. Can you please check the failed tests?

File: tests/helpers/dummy.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import shutil
+import tempfile
+
+import pytorch_lightning as pl
+import torch
+from omegaconf import OmegaConf
```
Comment (jpcbertoldo) [inline tests/helpers/dummy.py:6]:
```suggestion
```

File: tests/helpers/dummy.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    def val_dataloader(self) -> DataLoader:
+        return DataLoader(DummyDataset())
+
+    def test_dataloader(self) -> DataLoader:
+        return DataLoader(DummyDataset())
+
+
+class DummyModel(nn.Module):
+    def __init__(self):
+        super().__init__()
```
Comment (jpcbertoldo) [inline tests/helpers/dummy.py:34]:
```suggestion
        # pytlint: disable=useless-parent-delegation
        super().__init__()
```

File: tests/helpers/dummy.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        return DataLoader(DummyDataset())
+
+    def test_dataloader(self) -> DataLoader:
+        return DataLoader(DummyDataset())
+
+
+class DummyModel(nn.Module):
+    def __init__(self):
+        # pytlint: disable=useless-parent-delegation
+        super().__init__()
```
Comment (jpcbertoldo) [inline tests/helpers/dummy.py:35]:
```suggestion
class DummyModel(nn.Module):
    pass
```

**Comment by jpcbertoldo** [comment]
> > I believe i addressed everything.
> 
> Thanks @jpcbertoldo. Can you please check the failed tests?

Done.

@ashwinvaidya17 I forgot to mention that there was a check bothering about the fact that a few dummy classes where repeated in the tests so i put them in `tests/helpers/dummy.py`

**Comment by samet-akcay** [comment]
> > > I believe i addressed everything.
> > 
> > 
> > Thanks @jpcbertoldo. Can you please check the failed tests?
> 
> Done.
> 
> @ashwinvaidya17 I forgot to mention that there was a check bothering about the fact that a few dummy classes where repeated in the tests so i put them in `tests/helpers/dummy.py`

@jpcbertoldo, thanks for organising them. 
The tests needs another round of refactoring anyway, which we were planning on executing sometime this quarter. Would be a start.

**Comment by samet-akcay** [comment]
> > > I believe i addressed everything.
> > 
> > 
> > Thanks @jpcbertoldo. Can you please check the failed tests?
> 
> Done.

@jpcbertoldo, looks like the tests still fail unfortunately :/

**Comment by jpcbertoldo** [comment]
I dont understand, it was all green last night 🤨🤔. Ok will take a look.

**Comment by ashwinvaidya17** [comment]
> > > I believe i addressed everything.
> > 
> > 
> > Thanks @jpcbertoldo. Can you please check the failed tests?
> 
> Done.
> 
> @ashwinvaidya17 I forgot to mention that there was a check bothering about the fact that a few dummy classes where repeated in the tests so i put them in `tests/helpers/dummy.py`

Thanks for doing that. We have a lot of dummy modules and dataloaders peppered around the tests. That's something that needs to be addressed in the redesign as well.

File: anomalib/utils/metrics/__init__.py [hunk 6]
```diff
@@ -68,3 +51,125 @@ def metric_collection_from_names(metric_names: List[str], prefix: Optional[str])
...
+    cls = getattr(module, class_name)
+    return cls
+
+
+def metric_collection_from_dicts(metrics: Dict[str, Dict[str, Any]], prefix: Optional[str]) -> AnomalibMetricCollection:
+    """Create a metric collection from a dict of "metric name" -> "metric specifications".
+
+    Example:
+
+    metrics = {
```
Comment (ashwinvaidya17) [inline anomalib/utils/metrics/__init__.py:98]: Thanks for adding the example. I think you need to indent the section contents for sphinx to correctly format the contents.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks. One minor comment.

File: anomalib/utils/metrics/__init__.py (hunk 6 above)
Comment (jpcbertoldo) [reply anomalib/utils/metrics/__init__.py:98]: Sorry, I don't get it. Isn't it already indented?
Could you propose the change as a suggestion?

Comment (ashwinvaidya17) [reply anomalib/utils/metrics/__init__.py:98]: Maybe it is and github is formatting it incorrectly but I mean metrics and the other lines should not be on the same column as Example. 
It is currently 
```python
    Example:

    metrics = {
        "PixelWiseF1Score": {
            "class_path": "torchmetrics.F1Score",
            "init_args": {},
        },
```
and I mean

```python
    Example:

        metrics = {
            "PixelWiseF1Score": {
                "class_path": "torchmetrics.F1Score",
                "init_args": {},
            },
```
Not sure if this example makes it clear.

**Comment by jpcbertoldo** [comment]
> > > > I believe i addressed everything.
> > > 
> > > 
> > > Thanks @jpcbertoldo. Can you please check the failed tests?
> > 
> > 
> > Done.
> 
> @jpcbertoldo, looks like the tests still fail unfortunately :/

Ok, it was the test that had a little refactor, my bad!

```text
=========================== short test summary info ============================
ERROR tests/pre_merge/utils/callbacks/visualizer_callback/test_visualizer.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
```

I think it'lll fix.
I'm struggling to run tox locally because of some env var so let's see...

File: anomalib/utils/metrics/__init__.py (hunk 6 above)
Comment (jpcbertoldo) [reply anomalib/utils/metrics/__init__.py]: done @ashwinvaidya17

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks

**Review by samet-akcay — APPROVED** [review]
@jpcbertoldo, thanks looking good now!
